src/python/pants/engine/internals/parametrize.py

```diff
@@ -111,14 +156,26 @@ def expand(
                 {field_name: alias for field_name, alias, _ in parametrized_args}
             )
             parametrized_args_fields = tuple(
-                (field_name, field_value) for field_name, _, field_value in parametrized_args
+                (field_name, field_value)
+                for field_name, _, field_value in parametrized_args
+                if not isinstance(field_value, Parametrize)
             )
             expanded_fields: dict[str, Any] = dict(non_parametrized + parametrized_args_fields)
+            expanded_fields.update(
+                next(
+                    (
+                        group.kwargs
+                        for _, _, group in parametrized_args
+                        if isinstance(group, Parametrize)
+                    ),
+                    {},
+                )
+            )
```

Comment:
all parametrization groups are added as one term to the `parametrized` list of parametrizations to take the cartesian product of (on line 141):
```python
        if parametrized_groups:
            parametrized.append(parametrized_groups)
```

so when looping over the cross-product, there will ever only be at most one parametrized  group.

Reply:
there's a gotcha here to watch out for though, and that is we a user provides nested parametrizations.. unexpected and likely undefined behavior/unsupported, but we shouldn't act on it here as if it were a group. will address.

Reply:
Oh. I see the `itertools.product(...)` bit that I missed/misunderstood before. Now it makes sense.

> there's a gotcha here to watch out for though, and that is we a user provides nested parametrizations.. unexpected and likely undefined behavior/unsupported

You're talking about something like this, right?
```
target(
    parametrize("A", x=parametrize(1, 2, 3)),
    ...
)
```

If so, yes that would be really odd. Maybe we should eagerly fail if any of the args has a Parametrize object for a field value.

Reply:
> ```target(
    parametrize("A", x=parametrize(1, 2, 3)),
    ...
)
```

Ah, now I didn't consider that case.

Mine was for:
```
target(
    parametrize("A", x=0),
    x=parametrize(1, 2, 3),
    ...
)
```

Reply:
> If so, yes that would be really odd. Maybe we should eagerly fail if any of the args has a Parametrize object for a field value.

This is tricky, because you may need to `Parametrize.expand()` to get to the next level of field values.. and then you can take that and call `Parametrize.expand()` again, potentially. So we can't fail early really. Not exactly sure how `overrides` is implemented but it has support for parametrize in there..